src/poisson_fft.f90

```diff
@@ -0,0 +1,114 @@
+module m_poisson_fft
+   use m_allocator, only: field_t
+   use m_common, only: dp, pi
+   use m_tdsops, only: dirps_t
+
+   implicit none
+
+   type, abstract :: poisson_fft_t
+      !! FFT based Poisson solver
+      !! It can only handle 1D decompositions along z direction.
+      integer :: nx, ny, nz
+      complex(dp), allocatable, dimension(:, :, :) :: waves
+      complex(dp), allocatable, dimension(:) :: ax, bx, ay, by, az, bz
+   contains
+      procedure(fft_forward), deferred :: fft_forward
+      procedure(fft_backward), deferred :: fft_backward
+      procedure(fft_postprocess), deferred :: fft_postprocess
+      procedure :: base_init
+      procedure :: waves_set
+   end type poisson_fft_t
+
+   abstract interface
+      subroutine fft_forward(self, f_in)
+         import :: poisson_fft_t
+         import :: field_t
+         implicit none
+
+         class(poisson_fft_t) :: self
+         class(field_t), intent(in) :: f_in
+      end subroutine fft_forward
+
+      subroutine fft_backward(self, f_out)
+         import :: poisson_fft_t
+         import :: field_t
+         implicit none
+
+         class(poisson_fft_t) :: self
+         class(field_t), intent(inout) :: f_out
+      end subroutine fft_backward
+
+      subroutine fft_postprocess(self)
+         import :: poisson_fft_t
+         implicit none
+
+         class(poisson_fft_t) :: self
+      end subroutine fft_postprocess
+   end interface
+
+contains
+
+   subroutine base_init(self, xdirps, ydirps, zdirps)
+      implicit none
+
+      class(poisson_fft_t) :: self
+      class(dirps_t), intent(in) :: xdirps, ydirps, zdirps
+
+      integer :: nx, ny, nz
+
+      self%nx = xdirps%n; self%ny = ydirps%n; self%nz = zdirps%n; 
+
+      allocate (self%ax(self%nx), self%bx(self%nx))
+      allocate (self%ay(self%nx), self%by(self%nx))
+      allocate (self%az(self%nx), self%bz(self%nx))
+
+      allocate (self%waves(self%nx, self%ny, self%nz))
+
+      ! waves_set requires some of the preprocessed tdsops variables.
+      call self%waves_set(xdirps, ydirps, zdirps)
+
+   end subroutine base_init
+
+   subroutine waves_set(self, xdirps, ydirps, zdirps)
```

Comment:
Needs a high-level description; no idea what `waves_set` means.

Reply:
This subroutine is not finalised yet, I'll add a description later when its finished. I realised at this stage that it needs some stuff in dirps that we didn't bother storing at tdsops class level.